### PR TITLE
Refactor CompositeSPOSet

### PR DESCRIPTION
**Pull request by PhilipFackler**
## Proposed changes

Implement `CompositeSPOSetT<T>` to eventually replace `CompositeSPOSet`

## What type(s) of changes does this code introduce?

- Refactoring (no functional changes, no api changes)

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
Ubuntu 22.04

## Checklist

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by williamfgc** [comment]
@PhilipFackler something is off in the PR history, try to start from the current state of the branch

**Comment by PhilipFackler** [comment]
@williamfgc this one is unfinished. I missed several things. I'll let you know after I fix it.

**Comment by PhilipFackler** [comment]
@williamfgc should be right now.

File: src/QMCWaveFunctions/CompositeSPOSetT.h [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+	evaluate_notranspose(const ParticleSet& P, int first, int last,
+		ValueMatrix& logdet, GradMatrix& dlogdet,
+		HessMatrix& ddlogdet) override;
+	void
+	evaluate_notranspose(const ParticleSet& P, int first, int last,
+		ValueMatrix& logdet, GradMatrix& dlogdet, HessMatrix& ddlogdet,
+		GGGMatrix& dddlogdet) override;
+};
+
+// struct CompositeSPOSetBuilder : public SPOSetBuilder
```
Comment (williamfgc) [inline src/QMCWaveFunctions/CompositeSPOSetT.h]: Is this dead code in the original source? If so, we should remove it

File: src/QMCWaveFunctions/CompositeSPOSetT.cpp [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+		"evaluate_notranspose(P,first,last,logdet,dlogdet,ddlogdet,dddlogdet)");
+}
+
+// Class concrete types from ValueType
+template class CompositeSPOSetT<double>;
+template class CompositeSPOSetT<float>;
+template class CompositeSPOSetT<std::complex<double>>;
+template class CompositeSPOSetT<std::complex<float>>;
+
+// std::unique_ptr<SPOSetT<T>>
```
Comment (williamfgc) [inline src/QMCWaveFunctions/CompositeSPOSetT.cpp]: Similarly here, remove dead code.

**Review by williamfgc — CHANGES_REQUESTED** [review]

File: src/QMCWaveFunctions/CompositeSPOSetT.h (hunk 1 above)
Comment (PhilipFackler) [reply src/QMCWaveFunctions/CompositeSPOSetT.h]: @williamfgc, I was assuming we need to refactor the `*Builder` classes (there's one for `SPOSet` also), but that we are waiting for a later step to do that. Should I just remove it and reintroduce it later?

Comment (PhilipFackler) [reply src/QMCWaveFunctions/CompositeSPOSetT.h]: I guess to answer the question...it's not dead code, but code waiting to be refactored.

Comment (williamfgc) [reply src/QMCWaveFunctions/CompositeSPOSetT.h]: I see, thanks for the explanation. The goal is to be able to compile `CompositoteSPOSetT.cpp` as a single target.

**Comment by williamfgc** [comment]
@PhilipFackler is this ready?

**Pull request merged**